pkg/fleetcrt/fleetcrt.go

```diff
@@ -0,0 +1,91 @@
+// Package fleetcrt contains utilities to load and process TLS certificate files.
+package fleetcrt
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"os"
+)
+
+// Certificate holds a loaded certificate and its raw parts.
+type Certificate struct {
+	Crt    tls.Certificate
+	RawCrt []byte
+	RawKey []byte
+}
+
+// LoadCertificateFromFiles loads a TLS certificate from PEM cert and key file paths.
+//
+// Returns (nil, nil) if both files do not exist.
+func LoadCertificateFromFiles(crtPath, keyPath string) (*Certificate, error) {
+	checkFileExists := func(filePath string) (bool, error) {
+		switch s, err := os.Stat(filePath); {
+		case err == nil:
+			return !s.IsDir(), nil
+		case errors.Is(err, os.ErrNotExist):
+			return false, nil
+		default:
+			return false, err
+		}
+	}
+
+	crtExists, err := checkFileExists(crtPath)
+	if err != nil {
+		return nil, err
+	}
+	keyExists, err := checkFileExists(keyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if crtExists != keyExists {
+		return nil, fmt.Errorf(
+			"both crt and key files must exist: %s: %t, %s: %t",
+			crtPath, crtExists, keyPath, keyExists,
+		)
+	}
+	if !crtExists {
+		return nil, nil
+	}
+
+	crtBytes, err := os.ReadFile(crtPath)
+	if err != nil {
+		return nil, err
+	}
+	keyBytes, err := os.ReadFile(keyPath)
+	if err != nil {
+		return nil, err
+	}
+	crt, err := tls.X509KeyPair(crtBytes, keyBytes)
```

Comment:
FYI, the Go source code mentions this for `X509KeyPair`:
```
// X509KeyPair parses a public/private key pair from a pair of
// PEM encoded data. On successful return, Certificate.Leaf will be nil because
// the parsed form of the certificate is not retained.
```

And for the `Certificate.Leaf` field:

```
	// Leaf is the parsed form of the leaf certificate, which may be initialized
	// using x509.ParseCertificate to reduce per-handshake processing. If nil,
	// the leaf certificate will be parsed as needed.
```

So what we've done at startup for MDM-related certificates is this, to save on runtime handshake time: https://github.com/fleetdm/fleet/blob/main/server/config/config.go#L515-L522

Reply:
Oh cool!

Found the reasoning for the default behavior [here](https://groups.google.com/g/golang-dev/c/VResvFj2vF8):
> The Leaf member is only needed for clients doing client-authentication. This is rare compared to the common case of loading certificates for serving. In the latter case, the parsed form isn't needed because the server just sends the blob to the client and doesn't generally care what's in it.

Addressed in 935894089.

Reply:
Ain't that awesome when you have a crypto question and you get an answer from Adam Langley :D That's good to know!